pkg/activator/throttler.go

```diff
@@ -0,0 +1,177 @@
+}
+
+// NewThrottler creates a new Throttler
+func NewThrottler(params ThrottlerParams) *Throttler {
+	breakers := make(map[RevisionID]*queue.Breaker)
+	return &Throttler{breakers: breakers, breakerParams: params.BreakerParams, logger: params.Logger, getEndpoints: params.GetEndpoints, getRevision: params.GetRevision}
+}
+
+// Throttler keeps the mapping of Revisions to Breakers
+// and allows updating max concurrency dynamically of respective Breakers.
+// Max concurrency is essentially the number of semaphore tokens the Breaker has in rotation.
+// The manipulation of the parameter is done via `UpdateCapacity()` method.
+// It enables the use case to start with max concurrency set to 0 (no requests are sent because no endpoints are available)
+// and gradually increase its value depending on the external condition (e.g. new endpoints become available)
+type Throttler struct {
+	breakers      map[RevisionID]*queue.Breaker
+	breakerParams queue.BreakerParams
+	logger        *zap.SugaredLogger
+	getEndpoints  func(RevisionID) (int32, error)
+	getRevision   func(RevisionID) (*v1alpha12.Revision, bool, error)
+	mux           sync.Mutex
+}
+
+// Remove deletes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity updates the max concurrency of the Breaker corresponding to a revision.
+// It create a new breaker and saves into our bookkeeping if doesn't exist.
+// This is important for not loosing the update signals
+// for requests that come before the requests reached the Activator's Handler.
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	_, err := t.updateCapacity(rev, size)
+	return err
+}
+
+func (t *Throttler) updateCapacity(rev RevisionID, size int32) (breaker *queue.Breaker, err error) {
+	revision, exists, err := t.getRevision(rev)
+	if err != nil {
+		return breaker, err
+	}
+	if !exists {
+		return breaker, fmt.Errorf("no revision was found for: %s", rev)
+	}
+	cc := int32(revision.Spec.ContainerConcurrency)
+
+	if cc == 0 {
+		cc = t.breakerParams.MaxConcurrency
+	}
+	breaker, ok := t.breakers[rev]
+	if !ok {
+		breaker = queue.NewBreaker(t.breakerParams)
+		t.breakers[rev] = breaker
+	}
+	delta := size*cc - breaker.Capacity()
+	// Either the same number of endpoints received or
+	// some endpoints that were not registered at this Activator are torn down.
+	// In both cases do not perform the update.
+	if delta == 0 || (delta < 0 && breaker.Capacity() == 0) {
+		return breaker, nil
+	}
+	if err := breaker.UpdateConcurrency(delta); err != nil {
+		return breaker, err
+	}
+	return breaker, nil
+}
+
+// Try executes the `function` on the Breaker if it has enough capacity.
+// The function returns an error otherwise.
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	t.mux.Lock()
+	breaker, ok := t.breakers[rev]
```

Comment:
Currently in this code you have at least 2 return paths that won't unlock the lock.
This is easily solvable by
```
breaker, err := func () (*Breaker, error) {
 t.mux.Lock()
 defer t.mux.Unlock()
 ... rest of the code.
}()
if err != nil {return}
breaker.Maybe()
```

Reply:
Yeap, overseen this while adding new errors - good point